platform/xen/rumpxendev/evtdev.c

```diff
@@ -0,0 +1,502 @@
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/atomic.h>
+
+#include "rumpxen_xendev.h"
+
+#include <bmk-rumpuser/rumpuser.h>
+#include <bmk-core/memalloc.h>
+
+#pragma GCC diagnostic ignored "-Wcast-qual"
+/* mini-os/os.h has some bad casts */
+#include <mini-os/events.h>
+#include <mini-os/wait.h>
+#pragma GCC diagnostic error "-Wcast-qual"
+
+/* For ioctl interface. */
+#include "xenio3.h"
+
+/*----- data structures -----*/
+#define EVTDEV_RING_SIZE 2048
+#define EVTDEV_RING_MASK 2047
+#define BYTES_PER_PORT   4
+
+/* protect against multiple writer (multiple events triggers at the
+ * same time) there is only one reader, the worker thread.*/
+static kmutex_t xenevt_ring_lock;
+u_int xenevt_ring[EVTDEV_RING_SIZE];
+u_int xenevt_ring_prod, xenevt_ring_cons;
+
+struct xenevt_dev_data {
+	u_int ring[EVTDEV_RING_SIZE];
+	u_int ring_cons;
+	u_int ring_prod;
+#define EVTDEV_F_OVERFLOW 0x1 	/* ring overflow */
+	u_int flags;
+
+	kmutex_t lock;
+	kcondvar_t cv;
+	struct selinfo selinfo; /* used by poll, see select(9) */
+};
+
+/* Kernel event -> device instance mapping */
+static kmutex_t devevent_lock;
+static struct xenevt_dev_data *devevents[NR_EVENT_CHANNELS];
+
+/*----- helpers -----*/
+#define WBITS (POLLOUT | POLLWRNORM)
+#define RBITS (POLLIN  | POLLRDNORM)
+
+/* call with d->lock held */
+static void queue(struct xenevt_dev_data *d, u_int port)
+{
+	if (d->ring_cons == ((d->ring_prod + 1) & EVTDEV_RING_MASK)) {
+		d->flags |= EVTDEV_RING_MASK;
+		printf("evtdev: ring overflow port %d\n", port);
+	} else {
+		d->ring[d->ring_prod] = port;
+		membar_producer();
+		d->ring_prod = (d->ring_prod + 1) & EVTDEV_RING_MASK;
+	}
+	/* notify */
+	cv_signal(&d->cv);
+	selnotify(&d->selinfo, RBITS, NOTE_SUBMIT);
+}
+
+static void rump_evtdev_callback(u_int port)
+{
+	mutex_enter(&xenevt_ring_lock);
+
+	if (xenevt_ring_cons == ((xenevt_ring_prod + 1) & EVTDEV_RING_MASK)) {
+		minios_printk("xenevt driver ring overflowed!\n");
+	} else {
+		xenevt_ring[xenevt_ring_prod] = port;
+		membar_producer();
+		xenevt_ring_prod = (xenevt_ring_prod + 1) & EVTDEV_RING_MASK;
+	}
+
+	mutex_exit(&xenevt_ring_lock);
+
+	minios_wake_up(&minios_events_waitq);
+}
+
+static void xenevt_thread_func(void *ign)
+{
+	u_int prod = xenevt_ring_prod;
+	u_int cons;
+
+	/* give us a rump kernel context */
+	rumpuser__hyp.hyp_schedule();
+	rumpuser__hyp.hyp_lwproc_newlwp(0);
+	rumpuser__hyp.hyp_unschedule();
+
+	for (;;) {
+		minios_wait_event(minios_events_waitq, xenevt_ring_prod != prod);
+		prod = xenevt_ring_prod;
+		cons = xenevt_ring_cons;
+
+		membar_sync();
+
+		while (cons != prod) {
+			u_int port = xenevt_ring[cons];
+			struct xenevt_dev_data *d;
+
+			KASSERT(port < NR_EVENT_CHANNELS);
```

Comment:
I think this is OK. This is internal invariant -- port is queued by mini-os not user space, and we expect it to be less than NR_EVENT_CHANNELS.